src/gui/app.cpp

```diff
@@ -1,18 +1,150 @@
 #include "app.h"
 
+#include "arguments.h"
+#include "mainwindow.h"
+#include "printinfo.h"
+#include "version.h"
+
 #include <QDir>
 #include <QFileInfo>
 #include <QFileOpenEvent>
 #include <QSettings>
 #include <QTextStream>
-
-#include "arguments.h"
-#include "mainwindow.h"
-#include "printinfo.h"
-#include "version.h"
+#include <vector>
 
 namespace NeovimQt {
 
+MainWindow* s_lastActiveWindow{ nullptr };
+std::vector<MainWindow*> s_windows;
+int s_exitStatus{ 0 };
```

Comment:
Let's avoid globals whenever possible... Globals are tricky, and it is easy to accidentally introduce bugs.

You've also hit a few C++ gotchas here:

1. `s_` should only be used for `static` vars.
2. We should avoid globals with non-trival CTORs: `std::vector`.
3. Globals should be marked `static` to prevent this variable from being accessed outside of this compilation unit.

Can we define these variables somewhere else?

Reply:
I usually use the anonymous namespace for this. I think I got visually tricked by the namespace above, which I don't normally do, and just put the variables there. I can move them to the anonymous namespace if that's acceptable to you.

Reply:
No worries, C++ is tricky. It helps to have multiple pairs of eyes reading the code.

Here's my order of preference:
1. statics live within the function they are used
2. statics live within a class
3. statics live at the top of the compilation unit (separate namespace is a nice touch)

The other issues still apply inside a namespace:
1. These need to be `static` so the linker does not expose them outside of this file.
2. `std::vector` should be avoided as a global. The CTOR gets called from startup code, this can be (silently) expensive.

Reply:
> The other issues still apply inside a namespace:

I don't think those issues apply If I put them inside the unnamed namespace. Or are you saying this because you don't want me using the unnamed namespace?

> std::vector should be avoided as a global. The CTOR gets called from startup code, this can be (silently) expensive.

It`s simple container of pointers. This container is going to be used throughout the application's lifetime and it's useful as soon as it's launched. I don't think it would affect a lot, but I'm happy to change it if you can tell me your preferred alternative.

Reply:
> It`s simple container of pointers.

Is it a big performance hit? No.

However, it is a faux-paux in C++. You can very easily introduce expensive code at startup with a global `std::vector`. Please don't do this, let's avoid future headaches.

> I don't think those issues apply If I put them inside the unnamed namespace.

As I understand, yes the `static` linkage issues still apply.

Here is an example that avoids all of the issues above:
```
class App
{
...
static std::vector<const MainWindow&> m_windowList;
...
void AppendNewWindow(...) noexcept
{

}
void RemoveNewWindow(...) noexcept;
MainWindow* GetLastActiveWindow(...) noexcept;
}
```

Another alternative:
```
static std::vector<MainWindow&> GetWindowVectorList(...) noexcept;
{
    static std::vector<MainWIndow&> s_windowList{};
    // Do stuff?
    return s_windowList;
}
```

The compiler add some implicit code in both cases to initialize the vector on first use. Both are also easier to modify in the future.

Reply:
> Here's my order of preference:
> 
>   - statics live within the function they are used
>   - statics live within a class
>   - statics live at the top of the compilation unit (separate namespace is a nice touch)

That's a good rule of thumb. I'll keep that in mind. Thanks.

> You've also hit a few C++ gotchas here:
>     - s_ should only be used for static vars.

When put inside the anonymous name space, those variables are the same as statics. So, I don't see a
problem with the naming scheme here.

>     - We should avoid globals with non-trival CTORs: std::vector.

Correct, but in this case the lifetime of these variables are actually very tightly connected to the
life time of the application. If these were static class memebers, then the initialization order
would only change slightly. But I don't feel strongly about it, not that big of a difference. I'll
use your suggested solution.

>     - Globals should be marked static to prevent this variable from being accessed outside of this compilation unit.

Again, once they are in the anoynmous name space, they are static and only available to the current
complication unit.



>> I don't think those issues apply If I put them inside the unnamed namespace.
>
> As I understand, yes the static linkage issues still apply.

If you are talking about the static linkage issues you laid out above, I don't think they do. The
only time that difference might matter is when I put that static variable in a method, but then the
variable is used as soon as the application launches, I don't think it really doesn't make a
difference. But again, I don't feel strongly about them and it doesn't really make a mountain of
difference this way or the other. I'll make the changes.

Reply:
C++ is complicated, there are so many ways to solve the same problem.

Okay... I learned something new here about anonymous namespaces. Thanks! :smile:

Any solution works for me, as long as the linkage is static. However you think this code flows best.

Reply:
Feel free to keep `std::vector`, although I would recommend `MainWindow&` and `const` if possible.

There isn't a case where we expect a `nullptr`, so use a reference. References are better than the implicit never-null pointers with `assert` everywhere.